Documentation/design-docs/native-hosting-initialize.md

```diff
@@ -0,0 +1,182 @@
+  * Consider explicitly documenting types of behaviors which nobody should take dependency on (specifically failure scenarios)
+* Extensible
+
+## New scenarios
+The API should allow these, doesn't have to support them all in the beginning. It should be possible to ship with only some of these supported.
+* Runtime properties
+  * Ability to specify additional runtime properties from the native host
+  * Ability for the host to implement conflict resolution for runtime properties
+  * Ability for the host to inspect calculated runtime properties (the ones calculated by `hostfxr`/`hostpolicy`)
+
+
+## New `hostfxr` APIs
+
+### Initialize the host
+``` C++
+using hostfxr_handle = void *;
+
+struct hostfxr_initialize_parameters
+{
+    size_t size;
+    const char_t * host_path;   // path to the .exe
+    const char_t * dotnet_root  // resolved installation root (where fxr was found)
+};
+```
+
+The `hostfxr_initialize_parameters` structure stores initialize parameters which are common to all forms of initialization.
+* `size` - the size of the structure. This is used for versioning. Should be set to `sizeof(hostfxr_initialize_parameters)`.
+* `host_path` - path to the native host (typically the `.exe`). This value is not used for anything by the host, it's just passed to the CoreCLR as the path to the executable. It can point to a file which is not executable itself, if such file doesn't exist (for example in COM activation scenarios this points to the `comhost.dll`).
+* `dotnet_root` - path to the root of the .NET Core installation in use. This typically points to the install location from which the `hostfxr` has been loaded. Most commonly this would be (on Windows) `C:\Program Files\dotnet`. The path is used to search for shared frameworks and potentially SDKs.
+
+
+``` C++
+int hostfxr_initialize_for_app(
+    int argc,
+    const char_t * argv[],
+    const char_t * app_path  
+    const hostfxr_initialize_parameters * parameters,
+    hostfxr_handle * handle
+);
+```
+
+Initializes the host for running a managed application. Also used for running SDK commands.
+When used to execute an app, the `app_path` (or CLI equivalent) will be used to locate the `.runtimeconfig.json` and the `.deps.json` which will be used to load the application and its dependent frameworks.
+* `argc` and `argv` - the command line - entirely optional
+* `app_path` - path to the application (the managed `.dll`) to run. This can be `nullptr` if the app is specified in the command line arguments or if it's an SDK command.
+* `parameters` - additional parameters - see `hostfxr_initialize_parameters` for details. (Could be made optional potentially)
+* `handle` - output parameter. On success receives an opaque value which identifies the initialized host. The handle should be closed by calling `hostfxr_shutdown`.
+
+This function can only be called once per-process. It's not supported to run multiple apps in one process (even sequentially).
```

Comment:
It can be mixed - one app many components (runtime configs).
Yes - the function will fail if there already is runtime in the process. Running an app "takes over" the process in many ways - this is runtime behavior and we don't have a reason to change it now.

Reply:
I added statements to this effect in the document.